src/Workspaces/Core/Portable/Workspace/Solution/ProjectState_Checksum.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Serialization;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class ProjectState
+    {
+        public bool TryGetStateChecksums(out ProjectStateChecksums stateChecksums)
+        {
+            return _lazyChecksums.TryGetValue(out stateChecksums);
+        }
+
+        public async Task<Checksum> GetChecksumAsync(CancellationToken cancellationToken)
+        {
+            var collection = await _lazyChecksums.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            return collection.Checksum;
+        }
+
+        private async Task<ProjectStateChecksums> ComputeChecksumsAsync(CancellationToken cancellationToken)
+        {
+            // get states by id order to have deterministic checksum
+            var documentChecksumsTasks = DocumentIds.Select(id => DocumentStates[id].GetChecksumAsync(cancellationToken));
+            var additionalDocumentChecksumTasks = AdditionalDocumentIds.Select(id => AdditionalDocumentStates[id].GetChecksumAsync(cancellationToken));
+
+            var serializer = new Serializer(_solutionServices.Workspace.Services);
+
+            var infoChecksum = serializer.CreateChecksum(new SerializedProjectInfo(Id, Version, Name, AssemblyName, Language, FilePath, OutputFilePath, IsSubmission), cancellationToken);
+
+            var compilationOptionsChecksum = SupportsCompilation ? serializer.CreateChecksum(CompilationOptions, cancellationToken) : Checksum.Null;
+            var parseOptionsChecksum = SupportsCompilation ? serializer.CreateChecksum(ParseOptions, cancellationToken) : Checksum.Null;
+
+            var projectReferenceChecksums = new ProjectReferenceChecksumCollection(ProjectReferences.Select(r => serializer.CreateChecksum(r, cancellationToken)).ToArray());
```

Comment:
not understanding what you meant. you meant object[] ? once passed in, it is held as IReadOnlyList so can't be updated.